### PR TITLE
docs: add a guide for commit messages

### PR DESCRIPTION
**Pull request by lsf37**
Add a guide for how to write commit messages and commit message tags to make the messages more consistent and to help new people on the project get started more quickly.

### CODE REVIEW

**Comment by lsf37** [comment]
The list of currently used tags is generated from the repo and probably still contains some few that are inconsistent or unwanted. Help appreciated in weeding these out.

File: docs/commit-messages.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+arm ainvs
+arm aspec
+arm crefine
+arm haskell
+arm infoflow
+arm refine
+arm+arm-hyp crefine
+arm+arm-hyp machine
+arm+arm_hyp crefine
+arm+arm_hyp machine
```
Comment (lsf37) [inline docs/commit-messages.md:240]: the underscore version of `arm_hyp` is wrong

File: docs/commit-messages.md [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+arm-hyp refine
+asmrefine
+aspec
+aspec+access
+aspec+ainvs
+aspec+design+haskell
+aspec+haskell
+autocorres
+bisim
+c-kernel
```
Comment (lsf37) [inline docs/commit-messages.md:259]: do we want `c-kernel` or `ckernel`. Neither is a directory. The session is CKernel, so I'd be for `ckernel` as the tag.

File: docs/commit-messages.md [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+license
+misc
+misc/regression
+misc/scripts
+misc/stats
+proof
+proof+autocorres
+proof/Makefile
+proof/ROOT
+proofs
```
Comment (lsf37) [inline docs/commit-messages.md:328]: should probably only use `proof`, not `proofs`

File: docs/commit-messages.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+- > Where did this library lemma come from?
+
+## General
+
+The [seL4 repository guidelines][git-conventions] apply to the `l4v` repository,
+with the following exceptions and additions:
+
+- header can be max 78 characters
+- body is wrapped at 78 characters
+- we use tags in the header to in indicate which part of the repository
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:38]: Remove extra `in`

File: docs/commit-messages.md [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+For large directories that have logical sub parts, we use slashes, e.g.
+`lib/monads`. Not so much because the change is in that directory, but because
+we want to see that it's a library change and applies to the monad part of the
+library.
+
+If the change applies to many proofs, for instance large lemma renaming commits,
+we use tags such as `proof` and `spec`.
+
+We combine tags with `+` if a change applies to multiple parts, e.g.
+`clib+crefine` or `lib+proof`.
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:103]: `clib` is a subdirectory of `lib`, so maybe we should either use `lib/clib` or make an exception for `clib`. `clib` is used quite a few times below

File: docs/commit-messages.md [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+aarch64 aspec
+aarch64 aspec+ainvs
+aarch64 aspec+design
+aarch64 aspec+haskell
+aarch64 aspec+machine
+aarch64 cspec
+aarch64 design
+aarch64 design+machine
+aarch64 haskell
+aarch64 haskell+aspec
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:217]: This would go against the ordering you gave above, but maybe that doesn't matter too much

File: docs/commit-messages.md [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+arm+arm_hyp machine
+arm-hyp aspec
+arm-hyp aspec+design
+arm-hyp ainvs
+arm-hyp crefine
+arm-hyp design
+arm-hyp haskell
+arm-hyp haskell+refine
+arm-hyp machine
+arm-hyp proofs+ROOT
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:249]: `ROOT` seems like it should go before pretty much anything else, or be in its own commit, but I'm not sure

File: docs/commit-messages.md (hunk 2 above)
Comment (michaelmcinerney) [reply docs/commit-messages.md:259]: `ckernel` sounds good to me

File: docs/commit-messages.md [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+aspec+ainvs
+aspec+design+haskell
+aspec+haskell
+autocorres
+bisim
+c-kernel
+c-parser
+c-parser+autocorres
+c-parser+crefine
+c-parser+crefine+clib
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:263]: This again seems to violate the ordering

File: docs/commit-messages.md [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+riscv64+aarch64 ainvs
+run_tests
+run_tests+proof
+spec+proof
+style
+sys-init
+tools
+tools/asmrefine
+trivial
+word_lib
```
Comment (michaelmcinerney) [inline docs/commit-messages.md:367]: `word_lib` might be another exception to a subdirectory of `lib`

**Review by michaelmcinerney — APPROVED** [review]
Just some small nitpicks. But this looks very good and is very helpful. Especially for newcomers. Thanks Gerwin

**Comment by corlewis** [comment]
One thing I'm not sure about for the tags, are we more interested in the intention behind the commits or the specific files changed?
For example, let's say that we're making a change to `lib` and are then updating the rest of the proofs as required. However, it turns out that the only change needed is in `proof/refine/ARM`. In this case, which tag would be more useful, `proof` or `arm refine`?
I can see arguments for either, the latter is more useful for seeing which part of the repository have or have not changed, but it also implies more relevance than actually exists and could be distracting in the git history.

File: docs/commit-messages.md [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+## Header and Body
+
+There is good general advice on [commit message writing][commit-messages]
+available on the internet and it is as relevant to proofs as it is to code.
+You should read it, it's not long and it's good advice.
+
+Repeating some high-level points:
+
+- Use imperative voice (e.g. `proofs: rename lemma foo to bar`)
+- The header should be a short summary, focussing on "what"
```
Comment (corlewis) [inline docs/commit-messages.md:62]: It's debated but I think that `focusing` is more standard.

File: docs/commit-messages.md [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+  5-10 years from now. You can assume knowledge of the repository in general,
+  but you should not assume specific context that is obvious to you right now,
+  but that will not be known to a different person 5 years from now.
+
+## Tags
+
+- We use tags to indicate which part of the repository the commit applies to,
+  and if it is architecture-specific, then to which architecture it applies to.
+
+- We do no usually use branch tags, because git branches are ephemeral and we
```
Comment (corlewis) [inline docs/commit-messages.md:77]:
```suggestion
- We do not usually use branch tags, because git branches are ephemeral and we
```

File: docs/commit-messages.md [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+git log --stat     # additionally show which files have changed
+git log --p        # additionally show full diff
+```
+
+For all of these, you can supply a path to restrict the log to a certain file
+or directory in the repo. You can also supply a branch, or a commit range like
+`master..rt` to restrict the output.
+
+Esp `git log --oneline` is useful for quickly getting an overview. Example
+output:
```
Comment (corlewis) [inline docs/commit-messages.md:143]: This abbreviation is pretty common but it still feels better to me to use the full word
```suggestion
Especially `git log --oneline` is useful for quickly getting an overview. Example
output:
```

File: docs/commit-messages.md [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+b2dd5db6d lib/monads: fix document preparation issues
+03a045309 lib/monads: add AFP document setup
+d0bab9c79 misc/scripts: remove Darwin cpp wrapper
+```
+
+You can very quickly see that C verification has been active recently, that
+new tests were added, that AARCH64 proofs have been done, and that there was
+some work to do with the AFP and the monad library. You can see that nothing
+has happened with system initialiser or other user-level proofs, and that there
+are no changes that should affect for instance the C parser.
```
Comment (corlewis) [inline docs/commit-messages.md:170]:
```suggestion
has happened with the system initialiser or other user-level proofs, and that there
are no changes that should affect, for instance, the C parser.
```

**Comment by lsf37** [comment]
> One thing I'm not sure about for the tags, are we more interested in the intention behind the commits or the specific files changed? For example, let's say that we're making a change to `lib` and are then updating the rest of the proofs as required. However, it turns out that the only change needed is in `proof/refine/ARM`. In this case, which tag would be more useful, `proof` or `arm refine`? I can see arguments for either, the latter is more useful for seeing which part of the repository have or have not changed, but it also implies more relevance than actually exists and could be distracting in the git history.

Generally we want the intention. If it's cheap it's also good to be specific, though, because like you said it enables one to quickly see what has *not* changed. So in this specific case, I'd go for `arm refine`, but I'm open to arguments against.

File: docs/commit-messages.md (hunk 7 above)
Comment (lsf37) [reply docs/commit-messages.md:249]: I'd agree, `ROOT` should go first. If it makes a logical unit and is small enough I would not say it has to be separate commits, but depending on what it is, it probably often makes sense.

File: docs/commit-messages.md (hunk 6 above)
Comment (lsf37) [reply docs/commit-messages.md:217]: No, that's good, that's exactly the examples I was wanting to weed out. The list here is just what historically happened, the guide is supposed to be what we want to happen.

**Review by corlewis — APPROVED** [review]
Thanks for writing this up, it should be very helpful in us keeping a useful git history!

File: docs/commit-messages.md (hunk 5 above)
Comment (lsf37) [reply docs/commit-messages.md:103]: This is an interesting point and maybe we should discuss that one a bit. `CLib` is both a session in the global session name space and a `clib` is a subdirectory of `lib`. We don't say `proof/refine` etc either, so maybe we should just get rid of the slashes altogether. I've used them in the past for things that I thought too obscure/small for people to know or find quickly. I'd be open to just not doing that.

Comment (corlewis) [reply docs/commit-messages.md:103]: `Monads` is a case where I've been using slashes recently, for example `lib/monads/nondet`. `Monads` is also a global session, although a much smaller one, so I can live with or without the `lib` part of the tag. The suffix is useful here though, I often do want to disambiguate which monad I am working on and it's easier to just put it in the tag instead of trying to organically mention it in the header.

File: docs/commit-messages.md (hunk 10 above)
Comment (lsf37) [reply docs/commit-messages.md:62]: you're right, single `s` is the way to go. Feels weird to me, but maybe it's from `cussing` :-)

File: docs/commit-messages.md (hunk 3 above)
Comment (lsf37) [reply docs/commit-messages.md:322]: I was about to remove the `proofs` tag and standardise on `proof`, but now I'm not so sure any more. We (ok, mostly I) use that tag when many proofs are affected and it doesn't make sense to enumerate them all. `proofs` sounds better to me for that. Any opinions?

Comment (Xaphiosis) [reply docs/commit-messages.md:322]: I prefer `proof` because it's shorter, and in my mind stands for `proof update`, so I'd vote for `proof`. Note: if someone's mind thinks `update proofs` they'll come to the opposite conclusion.

File: docs/commit-messages.md (hunk 9 above)
Comment (Xaphiosis) [reply docs/commit-messages.md:361]: Right, word_lib is an odd one, because it's synced with AFP ... now that our monads are going upstream, they might hit similar naming awkwardness, since we'd want to mark what affects upstream.

File: docs/commit-messages.md (hunk 5 above)
Comment (Xaphiosis) [reply docs/commit-messages.md:103]: The slashes are visually a bit confusing, because they represent both paths and `or`. I guess in tags people shouldn't be saying `or`.
I'm fine with an exception for clib, even though it's physically placed under lib.

For consideration: as per my comment on `word_lib`, now that monads are getting upstreamed it might make sense to also make an exception for `monads` in that case. There is a bunch of work upcoming for concurrency, and so we'll be seeing `lib/monads/nondet` a fair bit, which could become `monads/nondet`.

File: docs/commit-messages.md [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+- Use imperative voice (e.g. `proofs: rename lemma foo to bar`)
+- The header should be a short summary, focusing on "what"
+- The body should explain what is going on in more detail (also in imperative
+  voice), but much more importantly *why* it is going on (is `bar` more
+  consistent than `foo`? Is the name `foo` needed for something else? Does `bar`
+  just better describe what is going on?).
+- You are trying to explain things to your future self or a future colleague
+  5-10 years from now. You can assume knowledge of the repository in general,
+  but you should not assume specific context that is obvious to you right now,
+  but that will not be known to a different person 5 years from now.
```
Comment (Xaphiosis) [inline docs/commit-messages.md:70]: I don't know if more emphasis is needed on not being lazy and/or doing header-only commits because they're "obvious". In theory yes, in practice would it really help ...

File: docs/commit-messages.md [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+`clib+crefine` or `lib+proof`.
+
+If something is specific to an architecture we preface the tag with the
+architecture, e.g. `arm refine:` or `aarch64 aspec+ainvs:`. The current
+architecture tags are: `arm`, `arm-hyp`, `x64`, `riscv`, `aarch64`.
+Please use these spellings only.
+
+More tag examples:
+
+- `trivial`: for small trivial changes like fixing a typo
```
Comment (Xaphiosis) [inline docs/commit-messages.md:112]: trivial has the extra role that no proofs/specs should be changed, i.e. a guarantee that a proof run will have the same result, so in theory it doesn't need to be regression tested (and in the past, regression tests would ignore testing these, not sure what's the situation now)

File: docs/commit-messages.md [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+output:
+
+```
+b3c6df48a clib: improve ccorres_While
+49ff8457f clib+crefine: improve and consolidate variants of ccorres_to_vcg
+8c433c085 clib: add some rules for hoarep
+82b954817 clib: improve ccorres_call_getter_setter
+8c59df449 lib/monads: remove more uses of _tac methods
+563232397 run_tests+proof: exclude SimplExportAndRefine for AARCH64
+1cce5b3ff proof: switch AArch64 quick_and_dirty from Refine to CRefine
```
Comment (Xaphiosis) [inline docs/commit-messages.md:152]: I think this one identifies are relevant distinction, because it's <inside the build system> <change an aarch64 setting>, as opposed to `aarch64: change something in the build system`, the latter I would want to be seen as incorrect

File: docs/commit-messages.md [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+4d0086567 lib/monads: add new Trace_* files to ROOT
+598e19dd6 lib/monads: coherent document structure
+4cbfb4ab0 lib/monads: minor style + warning cleanup
+b2dd5db6d lib/monads: fix document preparation issues
+03a045309 lib/monads: add AFP document setup
+d0bab9c79 misc/scripts: remove Darwin cpp wrapper
+```
+
+You can very quickly see that C verification has been active recently, that
+new tests were added, that AARCH64 proofs have been done, and that there was
```
Comment (Xaphiosis) [inline docs/commit-messages.md:167]: refine proofs, not all proofs :)

File: docs/commit-messages.md [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+
+If you're uncertain what tag to pick for a certain file or directory, the
+easiest way to figure it out is to do
+
+```sh
+git log --oneline <that-file>
+```
+
+Do your tags the same way people have done before. This will make the pattern
+consistent and should be reasonably easy to read even if it's not perfect.
```
Comment (Xaphiosis) [inline docs/commit-messages.md:186]: I'd also say if you're not sure, take a look at a few commits, because sometimes someone invented a new flavour in the most recent commit and you don't want to derail the train, but gently guide it back into balance.

File: docs/commit-messages.md [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+arm-hyp design
+arm-hyp haskell
+arm-hyp haskell+refine
+arm-hyp machine
+arm-hyp refine
+asmrefine
+aspec
+aspec+access
+aspec+ainvs
+aspec+design+haskell
```
Comment (Xaphiosis) [inline docs/commit-messages.md:251]: this points out historical order: `aspec < haskell/design < ainvs < refine < crefine`, with specs going before proofs, which I think should also be spelled out (did I miss that?)

Comment (Xaphiosis) [reply docs/commit-messages.md:251]: Oh yeah, I missed that. Might be worth expanding a little bit?

File: docs/commit-messages.md [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+
+For large directories that have logical sub parts, we use slashes, e.g.
+`lib/monads`. Not so much because the change is in that directory, but because
+we want to see that it's a library change and applies to the monad part of the
+library.
+
+If the change applies to many proofs, for instance large lemma renaming commits,
+we use tags such as `proof` and `spec`.
+
+We combine tags with `+` if a change applies to multiple parts, e.g.
```
Comment (Xaphiosis) [inline docs/commit-messages.md:102]: There's another thing to consider on the line below, when to *not* combine lib and proof changes. I sometimes want to `git show` the commit to see what was the relevant library change, and not see every update that could go into a `proof: update for <whatever lib changes>`.
This is not a clear-cut line, because it means there is a broken commit between the two, but it's very useful and our proofs can't really be bisected anyway due to the horrible runtime. I'm interested in people's thoughts on this.

File: docs/commit-messages.md [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+with the following exceptions and additions:
+
+- header can be max 78 characters
+- body is wrapped at 78 characters
+- we use tags in the header to indicate which part of the repository
+  the commit applies to
+
+We are using 78 for the header because of the tags, which take up some space. If
+you can manage to stay within 50 characters anyway, that is appreciated, but it's
+not always practical.
```
Comment (Xaphiosis) [inline docs/commit-messages.md:43]: we also don't kick people out for wrapping at 72 for the body, which is the default on many editors (including mine)

**Comment by Xaphiosis** [comment]
I think this is excellent, especially given the very short production timeframe (and no comma/just avalanche!
Added thoughts for discussion now that I had a sec, and I second Gerwin's opinion about intention and tagging above.

File: docs/commit-messages.md (hunk 5 above)
Comment (lsf37) [reply docs/commit-messages.md:103]: I'd be happy with that.

File: docs/commit-messages.md (hunk 20 above)
Comment (lsf37) [reply docs/commit-messages.md:102]: I'd leave this outside the commit message guide, because it's more a question of what commits one should make. If the proof upadte is small, I think combining them makes sense, but if the proof update is large I also like the split version better, because it enables you to figure out more easily what content has actually changed in the spec.

File: docs/commit-messages.md (hunk 14 above)
Comment (lsf37) [reply docs/commit-messages.md:70]: The paragraph is probably long enough. I don't think more emphasis would achieve much.

**Comment by lsf37** [comment]
Have addressed the feedback and am planning to merge when the tests are through. Still happy to discuss further changes, the content is not set in stone.
